File: CODE/powersum.f90
! power sum function
! 0 <= x <= n
! fstar = 0
!--------------------
!     POWER SUM FUNCTION
!--------------------
      subroutine funct(n,x,f)
      implicit none
      integer j,k,n
      real*8 f,x(n),s,b(4)
	b(1)=8.d0
	b(2)=18.d0
	b(3)=44.d0
	b(4)=114.d0
      f=0.d0
      do k=1,n
       s=0.d0
       do j=1,n
        s=s+x(j)**k
       end do
       f=f+(s-b(k))**2
      end do
      return
      end
!-------------
!     GRADIENT
!-------------
      subroutine grad(n,x,g)
      implicit none
      integer i,j,k,n
      real*8 x(n),g(n),s1,s,b(4)
	b(1)=8.d0
	b(2)=18.d0
	b(3)=44.d0
	b(4)=114.d0
      s1=0.d0
      do j=1,n
       s1=s1+x(j)
      end do
      do i=1,n
       g(i)=2.d0*(s1-b(1))
       do k=2,n
        s=0.d0
        do j=1,n
         s=s+x(j)**k
        end do
        g(i)=g(i)+2.d0*(s-b(k))*k*x(i)**(k-1)
       end do
      end do
      return
      end

File: Python/powersum.py
# power sum function
# 0 <= x <= n
# fstar = 0
#--------------------
#     POWER SUM FUNCTION
#--------------------


# TO TEST not known if it works

def funct(n, x):
    b = [8.0, 18.0, 44.0, 114.0]
    f = 0.0

    for k in range(1, n + 1):
        s = 0.0
        for j in range(1, n + 1):
            s += x[j - 1] ** k
        f += (s - b[k - 1]) ** 2

    return f


def grad(n, x):
    b = [8.0, 18.0, 44.0, 114.0]
    g = [0.0] * n

    s1 = sum(x)
    for i in range(1, n + 1):
        g[i - 1] = 2.0 * (s1 - b[0])
        for k in range(2, n + 1):
            s = 0.0
            for j in range(1, n + 1):
                s += x[j - 1] ** k
            g[i - 1] += 2.0 * (s - b[k - 1]) * k * x[i - 1] ** (k - 1)

    return g

if __name__ == "__main__":
    # Example Usage
    n_value = 4
    x_values = [1.0, 2.0, 3.0, 4.0]  # Replace with your actual input values

    result_funct = funct(n_value, x_values)
    result_grad = grad(n_value, x_values)

    print("Result (Function):", result_funct)
    print("Result (Gradient):", result_grad)
